src/workerd/api/global-scope.c++

```diff
@@ -524,28 +525,34 @@ void ServiceWorkerGlobalScope::sendHibernatableWebSocketMessage(
 }
 
 void ServiceWorkerGlobalScope::sendHibernatableWebSocketClose(
-    kj::String reason,
-    int code,
+    HibernatableSocketParams::Close close,
     Worker::Lock& lock, kj::Maybe<ExportedHandler&> exportedHandler) {
   auto event = jsg::alloc<HibernatableWebSocketEvent>();
 
   KJ_IF_MAYBE(h, exportedHandler) {
     KJ_IF_MAYBE(handler, h->webSocketClose) {
-      auto promise = (*handler)(lock, event->getWebSocket(lock), kj::mv(reason), code);
+      auto websocket = event->getWebSocket(lock);
+      websocket->initiateHibernatableRelease(lock, api::WebSocket::HibernatableReleaseState::CLOSE);
+      auto promise = (*handler)(lock, kj::mv(websocket), close.code, kj::mv(close.reason),
+                                close.wasClean);
       event->waitUntil(kj::mv(promise));
     }
     // We want to deliver close, but if no webSocketClose handler is exported, we shouldn't fail
   }
 }
 
 void ServiceWorkerGlobalScope::sendHibernatableWebSocketError(
+    kj::Exception e,
     Worker::Lock& lock,
     kj::Maybe<ExportedHandler&> exportedHandler) {
   auto event = jsg::alloc<HibernatableWebSocketEvent>();
 
   KJ_IF_MAYBE(h, exportedHandler) {
     KJ_IF_MAYBE(handler, h->webSocketError) {
-      auto promise = (*handler)(lock, event->getWebSocket(lock), event->getError(lock));
+      auto websocket = event->getWebSocket(lock);
+      websocket->initiateHibernatableRelease(lock, WebSocket::HibernatableReleaseState::ERROR);
+      jsg::Lock& js(lock);
+      auto promise = (*handler)(js, kj::mv(websocket), js.exceptionToJs(kj::mv(e)));
```

Comment:
@bcaimano this is where our js error becomes a jsg error, so unless it's tunneled it would just say internal error. That said, I want to come back to this after hibernation is finished and make both regular and hibernatable websocket errors a bit better (have a ticket internally)